Style engine: add border to frontend

ramonjd opened the pull request:
## What?
Adding border style generation to the style engine JS package.

Related backed PR: https://github.com/WordPress/gutenberg/pull/40531

## Why?
Carrying on with consolidating styles outlined in the tracking issue:
- https://github.com/WordPress/gutenberg/issues/38167

## How?

- Adding border style types and generator functions.
- Refactoring generateBoxRules to accommodate various "individual" properties

## Testing Instructions

I'm using Empty Theme to test.

Create a new post and add some blocks, such as Group or Column.

Update the border styles, and play around with individual border styles.

Here is some example block code:

<details>

<summary>Example HTML</summary>

```html
<!-- wp:group {"style":{"border":{"radius":{"topLeft":"29px","topRight":"60px","bottomLeft":"77px","bottomRight":"72px"},"top":{"color":"#b62323","style":"dashed","width":"61px"},"right":{"color":"#3c3737","width":"38px","style":"solid"},"bottom":{"color":"var:preset|color|pale-cyan-blue","width":"15.59rem"},"left":{"color":"#a61f76","width":"52px","style":"dotted"}}}} -->
<div class="wp-block-group" style="border-top-left-radius:29px;border-top-right-radius:60px;border-bottom-left-radius:77px;border-bottom-right-radius:72px;border-top-color:#b62323;border-top-style:dashed;border-top-width:61px;border-right-color:#3c3737;border-right-style:solid;border-right-width:38px;border-bottom-color:var(--wp--preset--color--pale-cyan-blue);border-bottom-width:15.59rem;border-left-color:#a61f76;border-left-style:dotted;border-left-width:52px"></div>
<!-- /wp:group -->

<!-- wp:group {"style":{"border":{"radius":"11px","style":"dashed","width":"20px"}},"borderColor":"vivid-green-cyan"} -->
<div class="wp-block-group has-border-color has-vivid-green-cyan-border-color" style="border-style:dashed;border-width:20px;border-radius:11px"></div>
<!-- /wp:group -->

<!-- wp:site-title {"style":{"color":{"background":"#cf1c1c"},"typography":{"fontStyle":"italic","fontWeight":"800","letterSpacing":"32px","lineHeight":3.9,"textTransform":"uppercase"},"border":{"radius":{"topLeft":"20px","topRight":"0px","bottomLeft":"20px","bottomRight":"100px"},"top":{"color":"var:preset|color|vivid-cyan-blue","width":"9px"},"right":{"color":"var:preset|color|luminous-vivid-orange","width":"18px"},"bottom":{"color":"var:preset|color|vivid-cyan-blue","width":"12px"},"left":{"color":"#b82b2b","width":"83px"}},"spacing":{"padding":{"top":"82px","right":"82px","bottom":"82px","left":"82px"},"margin":{"top":"67px","right":"67px","bottom":"67px","left":"67px"}}},"textColor":"pale-pink","fontSize":"x-large"} /-->

<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column {"style":{"border":{"top":{"color":"#863a3a","width":"41px"},"right":{"color":"var:preset|color|vivid-purple","width":"41px","style":"dashed"},"bottom":{"width":"57px"},"left":{"color":"var:preset|color|luminous-vivid-amber","width":"46px"}}}} -->
<div class="wp-block-column" style="border-top-color:#863a3a;border-top-width:41px;border-right-color:var(--wp--preset--color--vivid-purple);border-right-style:dashed;border-right-width:41px;border-bottom-width:57px;border-left-color:var(--wp--preset--color--luminous-vivid-amber);border-left-width:46px"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

</details>

Save the post.

Check that the editor and frontend match.

Run the tests!

```bash
npm run test-unit packages/style-engine/src/test/index.js
```

ramonjd committed:
Initial commit
Adding border style types and generator functions
Refactoring generateBoxRules to accommodate various "individual" properties

ramonjd commented on packages/style-engine/src/styles/utils.ts:
```diff
@@ -45,18 +51,20 @@ export function generateRule(
  *
  * @return GeneratedCSSRule[] CSS rules.
  */
 export function generateBoxRules(
```
As a follow up, maybe this could assume the responsibility of `generateRule`, similar to what  we do in the [backend](https://github.com/WordPress/gutenberg/blob/trunk/packages/style-engine/class-wp-style-engine.php#L346).

ramonjd commented on packages/style-engine/src/types.ts:
```diff
@@ -11,7 +11,31 @@ export type Box< T extends BoxVariants = undefined > = {
 
+export type BorderIndividualProperty = 'top' | 'right' | 'bottom' | 'left';
+export type BorderIndividualStyles< T extends BorderIndividualProperty > = {
+	color?: CSSProperties[ `border${ Capitalize< string & T > }Color` ];
```
My TS fu isn't great. Very glad to hear if there's a fancier way of doing this.

andrewserong replied:
My TS fu isn't very good, either, but this looks reasonable to me. I often find these rules a bit challenging to parse visually, but it given the sheer number of rules for each side, this reads better than the duplication of hard-coding each CSS property to me 👍

aaronrobertshaw replied:
Perhaps a simple inline comment with an example of the results might help people grok this quicker? Also, renaming `T` to something more explanatory could be another option.

ramonjd commented on packages/style-engine/src/styles/border/index.ts:
```diff
@@ -0,0 +1,150 @@
+const borderLeft = {
+	name: 'borderLeft',
+	generate: createBorderGenerateFunction( 'left' ),
+};
```
I'm starting to think that we could refactor the JS style engine to use properties from a master "definitions" constant, which is basically what we do in [hooks/style.js](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/style.js#L70) anyway using [__EXPERIMENTAL_STYLE_PROPERTY](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/constants.js#L17)

It's how we approached it in the [backend](https://github.com/WordPress/gutenberg/blob/trunk/packages/style-engine/class-wp-style-engine.php#L46).

And would allow us to store custom paths and CSS rules keys such as `border-$s-color` without the need to build them on the fly.

Not sure yet. I'll test it later.

andrewserong commented on packages/style-engine/src/styles/border/index.ts:
```diff
@@ -0,0 +1,150 @@
+	( style: Style, options: StyleOptions ) => {
+		const styleValue:
+			| BorderIndividualStyles< typeof individualProperty >
+			| undefined = get( style, [ 'border', individualProperty ] );
```
Does this need the lodash `get`, or can it be optional chaining?

E.g.

```suggestion
			| undefined = style?.border?.[ individualProperty ];
```

ramonjd replied:
Good call 👍

andrewserong commented on packages/style-engine/src/styles/border/index.ts:
```diff
@@ -0,0 +1,150 @@
+/**
+ * External dependencies
+ */
+import { camelCase, get } from 'lodash';
```
There's some active work underway to remove lodash from the repo (https://github.com/WordPress/gutenberg/issues/17025) — I'm not sure we can easily remove `camelCase` from the repo just yet, but there might be a simpler alternative for `get` (I've added a separate comment)

ramonjd replied:
> I'm not sure we can easily remove camelCase from the repo just yet

Maybe we could do it for the style engine though. We're also using `_.upperFirst()`. We could kill 2 birds with one method:

```js
const upperFirst = ( str ) => return ${ str.charAt(0).toUpperCase() + str.slice(1) };

const ruleKey = `border${ upperFirst( individualProperty ) }${ upperFirst( key ) }`;
```

I'll replace!

andrewserong replied:
Oh, I like it! And it makes sense for the style engine to have its own little utility methods like this 🙂

andrewserong reviewed (approved):
Nice one again @ramonjd! This LGTM and each of the individual border properties are testing well for me in the editor, with the same markup as prior to this PR 👍

Good question about whether to bring a single canonical list of the properties into the style engine package — it could be useful to centralise that knowledge, but it's still challenging for us to attempt to reduce the duplication between JS and PHP (I'm not sure if there's all that much we can do there for now 🤔). I suppose with the JS implementation, because things are split out into a separate style properties, we've got a bit more flexibility to have each module responsible for itself, whereas in PHP, we're a bit more bound by trying to keep it all into the one class, where a single list is very useful. At any rate, I agree it's something we could explore in follow-ups.

This change looks good to me! 🎉

ramonjd committed:
Added explainer comment for BorderIndividualStyles type
Removing lodash camelCase and upperFirst in favour of own implementation + test
Removing unnecessary lodash get()

mburridge commented:
Added the Needs Dev Note label in case this needs a dev note (either individual or as part of a "misc" dev note) for WP 6.1 release.
